### DISCUSSION TITLE
Best Practice for Multi Tenancy

### DISCUSSION THREAD

**Discussion by designermonkey** [authelia/authelia#5438] [General]
I want to be able to protect subdomains for my application, like `{tenant}.myapp.domain.com` for example.

What is the best advice so far for separating users per tenant?

Is it possible to load a different user file on a subdomain?

**Comment by james-d-elliott** [comment]
So the best solution in my opinion is to utilize the access control subjects criteria, you can then have users  in a single database and individually give them access to resources by username or by group membership.

Alternatively if you need separate files run multiple instances of authelia, one for each file, it's very light with less than ~20MB ram used and rarely more than 0% CPU.

**Comment by designermonkey** [comment]
Hi @james-d-elliott and thanks for the advice.

I see in the roadmap talk of multiple domain support. Is this similar to multi-tenancy?

I'll have a tinker around with your suggestions and see how I do.

**Reply by james-d-elliott** [reply]
Supporting multiple domains directly competes with what you're asking, it allows the reverse. `a.com` and `*.a.com` to share usernames and passwords with `b.com` and `*.b.com`. It already exists in master and works really well. I would also argue that multi-tenancy is a very vague term that can refer to at least 3 separate architectures these days.

I want to be clear about the following information, we're open to differing ideas and arguments in support of differing ideas, and open to changing our viewpoints.

If we were to support multi-tenancy the domains would have to be completely separate so they could be individually mapped to the cookie domains. For example `x.a.com` and `y.a.com` would *NOT* likely be intentionally supported as separate tenants, as they share a common root domain. 

The recommended and likely supported method would be by using ACL's as you can individually give users access to specific domains/subdomains./paths/etc. While technically there are ways to handle this the methods are fairly complex, and complexity in session management is a very clear security threat in my mind. There's just so many mistakes that could happen either in the code, how the user configures it, and how we validate that config.
